### DISCUSSION TITLE
.NET 8 Blazor Web App Migration - JavaScript

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3919] [General]
One of the items which has traditionally been challenging with Blazor is how to deal with JavaScript. 

In Interactive Blazor you needed to use JS Interop which meant creating C# wrappers for any JavaScript library or script you wanted to utilize, resulting in a more challenging integration story. In Static Blazor, JavaScript becomes a first-class citizen once again, which means you can include standard script references directly in your components, and they will function as expected.

This is a huge benefit for developers... but its a game changer for designers, as it now means that you can download or purchase any of the thousands of popular Bootstrap templates which are available online, and quickly integrate them into Oqtane. 

Bootstrap templates are comprised of HTML, CSS, and JavaScript... and although it was fairly easy to integrate the HTML and CSS previously, the JavaScript libraries were a challenge. Now you can simply reference the JavaScript libraries in your theme's razor components using standard ```<script />``` tags and bundle the resources with your theme.

I downloaded a Bootstrap theme today (https://bootstrapmade.com/demo/Arsha/) and created an Oqtane theme in ~2 hours which 100% replicates its content and behaviors (ie,. animations, etc...).

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/18dc1a0f-f805-46c7-bf40-57bd0e6c1ad3)

**Comment by sbwalker** [comment]
The steps I followed were:

- created a new theme using Theme Creator (Admin Dashboard / Theme Management / Create Theme)
- copied assets folder from BS theme to Client\wwwroot\Themes\Siliqon.Theme.Arsha
- added stylesheets to ThemeInfo.cs
- removed Settings types from ThemeInfo.cs and removed Settings razor components
- copied body content from BS theme to Default.razor
- added NavigationManager service
- replaced main content with ```<Pane Name="@PaneNames.Default" />```
- replaced src="assets/ with src="@ThemePath()assets/
- replaced logo markup with ```<a href="@NavigationManager.BaseUri" class="logo me-auto"><img src="@ThemePath()assets/img/logo.png" alt="@PageState.Site.Name"></a>```
- replaced with razor logic to render menu options
- added  ```<div class="mx-3 controls-group"><UserProfile /> <Login /> <ControlPanel /></div>```
- removed preloader div (this was causing some strange behavior when  navigating between pages)
- ran Oqtane and assigned the new theme to the Home page
- added an Html/Text module
- copied all section content from BS theme to the Html/Text module (fixing src attributes to point to src="Themes/Siliqon.Theme.Arsha/)

the Default.razor component below for reference:

```
@namespace Siliqon.Theme.Arsha
@inherits ThemeBase
@inject NavigationManager NavigationManager

  <!-- ======= Header ======= -->
  <header id="header" class="fixed-top ">
    <div class="container d-flex align-items-center">

      <a href="@NavigationManager.BaseUri" class="logo me-auto"><img src="@ThemePath()assets/img/logo.png" alt="@PageState.Site.Name"></a>

      <nav id="navbar" class="navbar">
        <ul>
                @foreach (var p in PageState.Pages.Where(item => item.IsNavigation))
                {
                    if (!p.Path.Contains("://"))
                    {
                        <li><a href="@NavigateUrl(p.Path)">@p.Name</a></li>
                    }
                    else
                    {
                        <li><a href="@p.Path">@p.Name</a></li>
                    }
                }
        </ul>
        <i class="bi bi-list mobile-nav-toggle"></i>
      </nav><!-- .navbar -->
        <div class="mx-3 controls-group"><UserProfile /> <Login /> <ControlPanel /></div>
    </div>
  </header><!-- End Header -->

  <!-- ======= Hero Section ======= -->
  <section id="hero" class="d-flex align-items-center">

    <div class="container">
      <div class="row">
        <div class="col-lg-6 d-flex flex-column justify-content-center pt-4 pt-lg-0 order-2 order-lg-1" data-aos="fade-up" data-aos-delay="200">
          <h1>Better Solutions For Your Business</h1>
          <h2>We are team of talented designers making websites with Bootstrap</h2>
          <div class="d-flex justify-content-center justify-content-lg-start">
            <a href="#about" class="btn-get-started scrollto">Get Started</a>
            <a href="https://www.youtube.com/watch?v=jDDaplaOz7Q" class="glightbox btn-watch-video"><i class="bi bi-play-circle"></i><span>Watch Video</span></a>
          </div>
        </div>
        <div class="col-lg-6 order-1 order-lg-2 hero-img" data-aos="zoom-in" data-aos-delay="200">
          <img src="@ThemePath()assets/img/hero-img.png" class="img-fluid animated" alt="">
        </div>
      </div>
    </div>

  </section><!-- End Hero -->
  
  <!-- ======= Main ======= -->
  <main id="main">
    <Pane Name="@PaneNames.Default" />
  </main><!-- End Main -->

  <!-- ======= Footer ======= -->
  <footer id="footer">

    <div class="footer-newsletter">
      <div class="container">
        <div class="row justify-content-center">
          <div class="col-lg-6">
            <h4>Join Our Newsletter</h4>
            <p>Tamen quem nulla quae legam multos aute sint culpa legam noster magna</p>
            <form action="" method="post">
              <input type="email" name="email"><input type="submit" value="Subscribe">
            </form>
          </div>
        </div>
      </div>
    </div>

    <div class="footer-top">
      <div class="container">
        <div class="row">

          <div class="col-lg-3 col-md-6 footer-contact">
            <h3>Arsha</h3>
            <p>
              A108 Adam Street <br>
              New York, NY 535022<br>
              United States <br><br>
              <strong>Phone:</strong> +1 5589 55488 55<br>
              <strong>Email:</strong> info@example.com<br>
            </p>
          </div>

          <div class="col-lg-3 col-md-6 footer-links">
            <h4>Useful Links</h4>
            <ul>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Home</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">About us</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Services</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Terms of service</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Privacy policy</a></li>
            </ul>
          </div>

          <div class="col-lg-3 col-md-6 footer-links">
            <h4>Our Services</h4>
            <ul>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Web Design</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Web Development</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Product Management</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Marketing</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="#">Graphic Design</a></li>
            </ul>
          </div>

          <div class="col-lg-3 col-md-6 footer-links">
            <h4>Our Social Networks</h4>
            <p>Cras fermentum odio eu feugiat lide par naso tierra videa magna derita valies</p>
            <div class="social-links mt-3">
              <a href="#" class="twitter"><i class="bx bxl-twitter"></i></a>
              <a href="#" class="facebook"><i class="bx bxl-facebook"></i></a>
              <a href="#" class="instagram"><i class="bx bxl-instagram"></i></a>
              <a href="#" class="google-plus"><i class="bx bxl-skype"></i></a>
              <a href="#" class="linkedin"><i class="bx bxl-linkedin"></i></a>
            </div>
          </div>

        </div>
      </div>
    </div>

    <div class="container footer-bottom clearfix">
      <div class="copyright">
        &copy; Copyright <strong><span>Arsha</span></strong>. All Rights Reserved
      </div>
      <div class="credits">
        <!-- All the links in the footer should remain intact. -->
        <!-- You can delete the links only if you purchased the pro version. -->
        <!-- Licensing information: https://bootstrapmade.com/license/ -->
        <!-- Purchase the pro version with working PHP/AJAX contact form: https://bootstrapmade.com/arsha-free-bootstrap-html-template-corporate/ -->
        Designed by <a href="https://bootstrapmade.com/">BootstrapMade</a>
      </div>
    </div>
  </footer><!-- End Footer -->

<a href="@NavigateUrl()" class="back-to-top d-flex align-items-center justify-content-center"><i class="bi bi-arrow-up-short"></i></a>

@if (PageState.RenderMode == RenderModes.Static)
{
    <!-- Vendor JS Files -->

    <script src="@ThemePath()assets/vendor/aos/aos.js"></script>
    <script src="@ThemePath()assets/vendor/bootstrap/js/bootstrap.bundle.min.js"></script>
    <script src="@ThemePath()assets/vendor/glightbox/js/glightbox.min.js"></script>
    <script src="@ThemePath()assets/vendor/isotope-layout/isotope.pkgd.min.js"></script>
    <script src="@ThemePath()assets/vendor/swiper/swiper-bundle.min.js"></script>
    <script src="@ThemePath()assets/vendor/waypoints/noframework.waypoints.js"></script>

    <!-- Template Main JS File -->
    <script src="@ThemePath()assets/js/main.js"></script>
}

@code {
    public override string Name => "Home";

    public override string Panes => PaneNames.Default;
}
```

and the ThemeInfo.cs:

```
using System.Collections.Generic;
using Oqtane.Models;
using Oqtane.Themes;
using Oqtane.Shared;

namespace Siliqon.Theme.Arsha
{
    public class ThemeInfo : ITheme
    {
        public Oqtane.Models.Theme Theme => new Oqtane.Models.Theme
        {
            Name = "Siliqon Arsha",
            Version = "1.0.0",
            PackageName = "Siliqon.Theme.Arsha",
            Resources = new List<Resource>()
            {
                // Google Fonts
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i%7CRaleway:300,300i,400,400i,500,500i,600,600i,700,700i%7CPoppins:300,300i,400,400i,500,500i,600,600i,700,700i" },
                // Vendor CSS Files
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/aos/aos.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/bootstrap/css/bootstrap.min.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/bootstrap-icons/bootstrap-icons.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/boxicons/css/boxicons.min.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/glightbox/css/glightbox.min.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/remixicon/remixicon.css" },
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/swiper/swiper-bundle.min.css" },
                // Template Main CSS File
                new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/css/style.css" },
                // JS file for Interactive rendering
                new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/bootstrap/js/bootstrap.bundle.min.js", Location = ResourceLocation.Body, RenderMode = RenderModes.Interactive }
           }
        };

    }
}
```

This was a very quick integration. If I had more time I would either include a ThemeSetting component or add additional Panes so that the hardcoded content would be editable and dynamic (ie. the Hero section and Footer). 

Also note that this theme would not function on Interactive render mode, as the ```<script />``` elements are not supported.

**Comment by sbwalker** [comment]
After more testing with the Arsha theme I realized that in order for a theme to be able to support both Static and Interactive render modes it was necessary to include an additional property on the Resource class for RenderMode. This allows a theme to include ```<script``` tags directly in the component for Static rendering scenarios as well as specify the scripts in the Resources collection for Interactive render modes. Without this new RenderMode property the framework would inject the scripts twice when running on Static render mode.

**Reply by techravi007** [reply]
Hi @sbwalker 
Can you pl share the git repo url for this ?

**Reply by sbwalker** [reply]
there is not a Git repo for this - I created an example project to test the 8.1 release

**Reply by techravi007** [reply]
Okay

**Comment by sbwalker** [comment]
Perhaps I was overly optimistic... this thread seems to indicate that including scripts within components does not work well with enhanced navigation:

https://github.com/dotnet/aspnetcore/issues/51331

"Using enhanced navigation requires a different approach when it comes down to dealing with scripts. The main reason being that without enhanced navigation, all the state is cleared on each navigation, while with it, the state on the page is "preserved". That means that scripts that modify global state or depend on the page loading to perform an action need to be adjusted to account for the fact that with enhanced nav, those two things don't work in the same way."

Basically this means that as users navigate between different pages in your Blazor app, you will encounter issues with any JavaScript which is expected to execute on every page. This is due to enhanced navigation which uses the browsers fetch API to retrieve a new page, then patches the DOM with the parts of the page that have changed. This avoids performing full page loads every time you navigate between pages (especially useful if only a small part of the UI has actually changed) - however it is not compatible with traditional JavaScript usage.

**Comment by sbwalker** [comment]
I found a solution which works... it is based on this sample project from Mackinnon Buck who is a member of the Blazor team at Microsoft:

https://github.com/MackinnonBuck/blazor-page-script

Blazor has a concept known as "JavaScript Initializers" (https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/startup?view=aspnetcore-8.0). They allow you to execute logic before and after a Blazor app loads. I added a JavaScript Initializer to the wwwroot named Oqtane.Server.lib.module.js (as this is the required naming convention). The initializer contains the JavaScript code from the blazor-page-script repo above. 

Essentially the script registers a custom HTML element called "page-script" which accepts a single parameter of "src". The purpose of the page-script element is to trigger events when a user navigates your application. The most critical event when running on Static Blazor using Enhanced Navigation (the default) is the "onUpdate" event. This event is called on every page transition and basically replaces the traditional role of the JavaScript "onload" event. 

A new property was added to the Oqtane Resource class called Reload. It defaults to False but when it is set to True it instructs the framework to inject the script into the page using the page-script custom element rather than a traditional script element. You would only use this new property for JavaScript files which need to be reloaded on every page transition.

For the Bootstrap theme I used as an example, I moved all of the script resources to the ThemeInfo.cs and declared them as Resources:

```
Resources = new List<Resource>()
{
      // Google Fonts
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i%7CRaleway:300,300i,400,400i,500,500i,600,600i,700,700i%7CPoppins:300,300i,400,400i,500,500i,600,600i,700,700i" },
      // Vendor CSS Files
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/aos/aos.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/bootstrap/css/bootstrap.min.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/bootstrap-icons/bootstrap-icons.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/boxicons/css/boxicons.min.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/glightbox/css/glightbox.min.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/remixicon/remixicon.css" },
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/vendor/swiper/swiper-bundle.min.css" },
      // Template Main CSS File
      new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/assets/css/style.css" },
      // Vendor JS Files
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/purecounter/purecounter_vanilla.js", Location = ResourceLocation.Body },
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/aos/aos.js", Location = ResourceLocation.Body },
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/bootstrap/js/bootstrap.bundle.min.js", Location = ResourceLocation.Body },
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/glightbox/js/glightbox.min.js", Location = ResourceLocation.Body },
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/isotope-layout/isotope.pkgd.min.js", Location = ResourceLocation.Body },
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/vendor/swiper/swiper-bundle.min.js", Location = ResourceLocation.Body },
      // Template Main JS File
      new Resource { ResourceType = ResourceType.Script, Url = "~/assets/js/main.js", Location = ResourceLocation.Body, Reload = true }
}
```

You will notice that the last script in the list, main.js, uses the new Reload = true property. This script file had to be modified to conform to the pattern required by the page-script element. It needed to implement the onUpdate event, and any references to window.addEventListener('load', xxx) needed to be removed (as "onload" is not triggered for page transitions).

```
function initializeTheme() {
  "use strict";

  // implementation
}

export function onUpdate() {
    initializeTheme();
}
```

Once these changes were made, the JavaScript works as expected even as you navigate to different page in the site... and all animations, etc... work perfectly.

**Reply by vnetonline** [reply]
@sbwalker, after making this change, did the preloader work?
